xhu.py

```diff
@@ -135,6 +179,11 @@ def put_file(path):
     )
 
     dest_path.parent.mkdir(parents=True, exist_ok=True, mode=0o770)
+
+    quota = flask.request.args.get("q", "")
+    if (quota):
+        apply_quota(dest_path.parent.parent, int(quota))
```

Comment:
Don’t make assumptions about the directory structure.

This should also take into account the size of the uploaded file. The amount of data used must still be below the quota even after the upload.

It also needs to handle the case where the file is larger than the quota allows for: in that case, the upload should be refused without any files being deleted.

Reply:
> This should also take into account the size of the uploaded file. The amount of data used must still be below the quota even after the upload.

> It also needs to handle the case where the file is larger than the quota allows for: in that case, the upload should be refused without any files being deleted.

Depends on whether you're enforcing a hard quota or a soft quota. I'm OK with usage slightly going over the quota. Prosody already enforces file size via `http_upload_external_file_size_limit`.

Reply:
> Don’t make assumptions about the directory structure.

I don't yet see how I can do it otherwise. Putting storage usage in a file like you suggested provides some decoupling, but you still need to know what files to remove.